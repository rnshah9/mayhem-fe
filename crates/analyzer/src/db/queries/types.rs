use std::rc::Rc;

use smol_str::SmolStr;

use crate::context::{AnalyzerContext, TempContext};
use crate::db::Analysis;
use crate::errors::TypeError;
use crate::namespace::items::{FunctionSigId, ImplId, TraitId, TypeAliasId};
use crate::namespace::scopes::ItemScope;
use crate::namespace::types::{self, TypeId};
use crate::traversal::types::type_desc;
use crate::AnalyzerDb;

/// Returns all `impl` for the given type from the current ingot as well as dependency ingots
pub fn all_impls(db: &dyn AnalyzerDb, ty: TypeId) -> Rc<[ImplId]> {
    let ingot_modules = db
        .root_ingot()
        .all_modules(db)
        .iter()
        .flat_map(|module_id| module_id.all_impls(db).to_vec())
        .collect::<Vec<_>>();

    db.ingot_external_ingots(db.root_ingot())
        .values()
        .flat_map(|ingot| ingot.all_modules(db).to_vec())
        .flat_map(|module_id| module_id.all_impls(db).to_vec())
        .chain(ingot_modules)
        .filter(|val| val.receiver(db) == ty)
        .collect()
}

pub fn impl_for(db: &dyn AnalyzerDb, ty: TypeId, treit: TraitId) -> Option<ImplId> {
    db.all_impls(ty)
        .iter()
        .find(|impl_| impl_.trait_id(db) == treit)
        .cloned()
}

pub fn function_sigs(db: &dyn AnalyzerDb, ty: TypeId, name: SmolStr) -> Rc<[FunctionSigId]> {
    ty.get_all_impls(db)
        .iter()
        .filter_map(|impl_| impl_.function(db, &name))
        .map(|fun| fun.sig(db))
        .chain(ty.function_sig(db, &name).map_or(vec![], |fun| vec![fun]))
        .collect()
}

pub fn type_alias_type(
    db: &dyn AnalyzerDb,
    alias: TypeAliasId,
) -> Analysis<Result<types::TypeId, TypeError>> {
    let mut scope = ItemScope::new(db, alias.data(db).module);
    let typ = type_desc(&mut scope, &alias.data(db).ast.kind.typ);

    Analysis::new(typ, scope.diagnostics.take().into())
}

pub fn type_alias_type_cycle(
    db: &dyn AnalyzerDb,
    _cycle: &[String],
    alias: &TypeAliasId,
) -> Analysis<Result<types::TypeId, TypeError>> {
    let mut context = TempContext::default();
    let err = Err(TypeError::new(context.error(
        "recursive type definition",
        alias.data(db).ast.span,
        "",
    )));

    Analysis::new(err, context.diagnostics.take().into())
}
