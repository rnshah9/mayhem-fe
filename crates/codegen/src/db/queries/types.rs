use fe_mir::ir::{
    types::{ArrayDef, EventDef, MapDef, StructDef, TupleDef},
    Type, TypeId, TypeKind,
};

use crate::db::CodegenDb;

pub fn legalized_type(db: &dyn CodegenDb, ty: TypeId) -> TypeId {
    let ty_data = ty.data(db.upcast());
    let ty_kind = match &ty.data(db.upcast()).kind {
        TypeKind::Tuple(def) => {
            let items = def
                .items
                .iter()
                .filter_map(|item| {
                    if item.is_zero_sized(db.upcast()) {
                        None
                    } else {
                        Some(legalized_type(db, *item))
                    }
                })
                .collect();
            let new_def = TupleDef { items };
            TypeKind::Tuple(new_def)
        }

        TypeKind::Array(def) => {
            let new_def = ArrayDef {
                elem_ty: legalized_type(db, def.elem_ty),
                len: def.len,
            };
            TypeKind::Array(new_def)
        }

        TypeKind::Struct(def) => {
            let fields = def
                .fields
                .iter()
                .cloned()
                .filter_map(|(name, ty)| {
                    if ty.is_zero_sized(db.upcast()) {
                        None
                    } else {
                        Some((name, legalized_type(db, ty)))
                    }
                })
                .collect();
            let new_def = StructDef {
                name: def.name.clone(),
                fields,
                span: def.span,
                module_id: def.module_id,
            };
            TypeKind::Struct(new_def)
        }

        TypeKind::Event(def) => {
            let fields = def
                .fields
                .iter()
                .cloned()
                .filter_map(|(name, ty, indexed)| {
                    if ty.is_zero_sized(db.upcast()) {
                        None
                    } else {
                        Some((name, legalized_type(db, ty), indexed))
                    }
                })
                .collect();
            let new_def = EventDef {
                name: def.name.clone(),
                fields,
                span: def.span,
                module_id: def.module_id,
            };
            TypeKind::Event(new_def)
        }

        TypeKind::Contract(def) => {
            let fields = def
                .fields
                .iter()
                .cloned()
                .filter_map(|(name, ty)| {
                    if ty.is_zero_sized(db.upcast()) {
                        None
                    } else {
                        Some((name, legalized_type(db, ty)))
                    }
                })
                .collect();
            let new_def = StructDef {
                name: def.name.clone(),
                fields,
                span: def.span,
                module_id: def.module_id,
            };
            TypeKind::Contract(new_def)
        }

        TypeKind::Map(def) => {
            let new_def = MapDef {
                key_ty: legalized_type(db, def.key_ty),
                value_ty: legalized_type(db, def.value_ty),
            };
            TypeKind::Map(new_def)
        }

        TypeKind::MPtr(ty) => {
            let new_ty = legalized_type(db, *ty);
            TypeKind::MPtr(new_ty)
        }

        TypeKind::SPtr(ty) => {
            let new_ty = legalized_type(db, *ty);
            TypeKind::SPtr(new_ty)
        }

        _ => return ty,
    };

    let analyzer_ty = ty_data.analyzer_ty;
    db.mir_intern_type(Type::new(ty_kind, analyzer_ty).into())
}
