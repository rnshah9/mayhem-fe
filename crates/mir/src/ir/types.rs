use fe_analyzer::namespace::items as analyzer_items;
use fe_analyzer::namespace::types as analyzer_types;
use fe_common::{impl_intern_key, Span};
use smol_str::SmolStr;

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct Type {
    pub kind: TypeKind,
    pub analyzer_ty: Option<analyzer_types::TypeId>,
}

impl Type {
    pub fn new(kind: TypeKind, analyzer_ty: Option<analyzer_types::TypeId>) -> Self {
        Self { kind, analyzer_ty }
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub enum TypeKind {
    I8,
    I16,
    I32,
    I64,
    I128,
    I256,
    U8,
    U16,
    U32,
    U64,
    U128,
    U256,
    Bool,
    Address,
    Unit,
    Array(ArrayDef),
    // TODO: we should consider whether we really need `String` type.
    String(usize),
    Tuple(TupleDef),
    Struct(StructDef),
    Event(EventDef),
    Contract(StructDef),
    Map(MapDef),
    MPtr(TypeId),
    SPtr(TypeId),
}

/// An interned Id for [`ArrayDef`].
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct TypeId(pub u32);
impl_intern_key!(TypeId);

/// A static array type definition.
#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct ArrayDef {
    pub elem_ty: TypeId,
    pub len: usize,
}

/// A tuple type definition.
#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct TupleDef {
    pub items: Vec<TypeId>,
}

/// A user defined struct type definition.
#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct StructDef {
    pub name: SmolStr,
    pub fields: Vec<(SmolStr, TypeId)>,
    pub span: Span,
    pub module_id: analyzer_items::ModuleId,
}

/// A user defined struct type definition.
#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct EventDef {
    pub name: SmolStr,
    pub fields: Vec<(SmolStr, TypeId, bool)>,
    pub span: Span,
    pub module_id: analyzer_items::ModuleId,
}

/// A map type definition.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct MapDef {
    pub key_ty: TypeId,
    pub value_ty: TypeId,
}
