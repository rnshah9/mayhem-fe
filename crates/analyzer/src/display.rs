use crate::AnalyzerDb;
use std::fmt;

pub trait Displayable: DisplayWithDb {
    fn display<'a, 'b>(&'a self, db: &'b dyn AnalyzerDb) -> DisplayableWrapper<'b, &'a Self> {
        DisplayableWrapper::new(db, self)
    }
}
impl<T: DisplayWithDb> Displayable for T {}

pub trait DisplayWithDb {
    fn format(&self, db: &dyn AnalyzerDb, f: &mut fmt::Formatter<'_>) -> fmt::Result;
}

impl<T: ?Sized> DisplayWithDb for &T
where
    T: DisplayWithDb,
{
    fn format(&self, db: &dyn AnalyzerDb, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        (*self).format(db, f)
    }
}

pub struct DisplayableWrapper<'a, T> {
    db: &'a dyn AnalyzerDb,
    inner: T,
}
impl<'a, T> DisplayableWrapper<'a, T> {
    pub fn new(db: &'a dyn AnalyzerDb, inner: T) -> Self {
        Self { db, inner }
    }
    pub fn child(&self, inner: T) -> Self {
        Self { db: self.db, inner }
    }
}

impl<T: DisplayWithDb> fmt::Display for DisplayableWrapper<'_, T> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        self.inner.format(self.db, f)
    }
}
