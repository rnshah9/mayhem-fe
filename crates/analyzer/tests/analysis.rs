use fe_analyzer::display::Displayable;
use fe_analyzer::namespace::items::{self, IngotId, IngotMode, Item, ModuleId, TypeDef};
use fe_analyzer::namespace::types::TypeId;
use fe_analyzer::{AnalyzerDb, TestDb};
use fe_common::diagnostics::{diagnostics_string, print_diagnostics, Diagnostic, Label, Severity};
use fe_common::files::{FileKind, Utf8Path};
use fe_parser::node::{NodeId, Span};
use indexmap::{indexmap, IndexMap};
use insta::assert_snapshot;
use smallvec::SmallVec;
use std::collections::{HashMap, VecDeque};
use std::fmt::Display;
use wasm_bindgen_test::wasm_bindgen_test;

#[test]
fn analyze_std_lib() {
    let mut db = TestDb::default();

    // Should return the same id
    let std_ingot = IngotId::std_lib(&mut db);
    let std_ingot_2 = IngotId::std_lib(&mut db);
    assert_eq!(std_ingot, std_ingot_2);

    let diags = std_ingot.diagnostics(&db);
    if !diags.is_empty() {
        print_diagnostics(&db, &diags);
        panic!("std lib analysis failed")
    }
}

#[test]
fn ingot_files_to_modules() {
    let mut db = TestDb::default();
    let ingot = IngotId::from_files(
        &mut db,
        "libcool",
        IngotMode::Lib,
        FileKind::Local,
        &[
            ("foo/fee.fe", ""),
            ("lib.fe", ""),
            ("bar/baz.fe", ""),
            ("bar/bum.fe", ""),
            ("x.fe", ""),
            ("a/b/c/d.fe", ""),
        ],
        indexmap! {},
    );

    let submod_map = |module: ModuleId| {
        module
            .submodules(&db)
            .iter()
            .map(|m| (m.name(&db), *m))
            .collect::<HashMap<_, _>>()
    };

    let lib = ingot.root_module(&db).unwrap();
    let subs = submod_map(lib);
    assert_eq!(subs.len(), 4);

    let bar_subs = submod_map(subs["bar"]);
    assert_eq!(bar_subs.len(), 2);
    assert!(bar_subs.contains_key("bum"));
    assert!(bar_subs.contains_key("baz"));

    let d = subs["a"].submodules(&db)[0] // b
        .submodules(&db)[0] // c
        .submodules(&db)[0]; // d
    assert_eq!(d.name(&db), "d");
}

macro_rules! test_analysis {
    ($name:ident, $path:expr) => {
        #[test]
        #[wasm_bindgen_test]
        fn $name() {
            let mut db = TestDb::default();

            let file_name = Utf8Path::new($path).file_name().unwrap();
            let module = ModuleId::new_standalone(&mut db, file_name, test_files::fixture($path));

            let diags = module.diagnostics(&db);
            if !diags.is_empty() {
                print_diagnostics(&db, &diags);
                panic!("analysis failed")
            }

            if cfg!(target_arch = "wasm32") {
                // NOTE: If this assertion fails, the generation of the output diff
                //  is very slow on wasm, and may result in an out-of-memory error
                //  for larger diffs. I recommend commenting out all tests but one.
                fe_common::assert_snapshot_wasm!(
                    concat!("snapshots/analysis__", stringify!($name), ".snap"),
                    build_snapshot(&db, module)
                );
            } else {
                assert_snapshot!(build_snapshot(&db, module));
            }
        }
    };
}

macro_rules! test_analysis_ingot {
    ($name:ident, $path:expr) => {
        #[test]
        #[wasm_bindgen_test]
        fn $name() {
            let mut db = TestDb::default();

            let std = IngotId::std_lib(&mut db);
            let ingot = IngotId::from_files(
                &mut db,
                "test_ingot",
                IngotMode::Main,
                FileKind::Local,
                &test_files::fixture_dir_files($path),
                indexmap! { "std".into() => std },
            );

            let diags = ingot.diagnostics(&db);
            if !diags.is_empty() {
                print_diagnostics(&db, &diags);
                panic!("analysis failed")
            }
            diagnostics_string(&db, &diags);

            let snapshot =
                ModuleIter::new(&db, ingot.root_module(&db).expect("missing root module"))
                    .map(|module| build_snapshot(&db, module))
                    .collect::<Vec<_>>()
                    .join("\n");

            if cfg!(target_arch = "wasm32") {
                // NOTE: If this assertion fails, the generation of the output diff
                //  is very slow on wasm, and may result in an out-of-memory error
                //  for larger diffs. I recommend commenting out all tests but one.
                fe_common::assert_snapshot_wasm!(
                    concat!("snapshots/analysis__", stringify!($name), ".snap"),
                    snapshot
                );
            } else {
                assert_snapshot!(snapshot);
            }
        }
    };
}

/// Breadth-first walk of a module tree
struct ModuleIter<'a> {
    db: &'a dyn AnalyzerDb,
    emit: VecDeque<ModuleId>,
    walk: VecDeque<ModuleId>,
}

impl<'a> ModuleIter<'a> {
    pub fn new(db: &'a dyn AnalyzerDb, root: ModuleId) -> Self {
        let mut deq = VecDeque::new();
        deq.push_back(root);
        Self {
            db,
            emit: deq.clone(),
            walk: deq,
        }
    }
}

impl<'a> Iterator for ModuleIter<'a> {
    type Item = ModuleId;
    fn next(&mut self) -> Option<ModuleId> {
        if let Some(modid) = self.emit.pop_front() {
            Some(modid)
        } else if let Some(modid) = self.walk.pop_front() {
            let submods = modid.submodules(self.db);
            self.walk.extend(submods.iter());
            self.emit = VecDeque::from_iter(submods.iter().copied());
            self.next()
        } else {
            None
        }
    }
}

test_analysis! { erc20_token, "demos/erc20_token.fe"}
test_analysis! { guest_book, "demos/guest_book.fe"}
test_analysis! { simple_open_auction, "demos/simple_open_auction.fe"}
test_analysis! { uniswap, "demos/uniswap.fe"}
test_analysis! { address_bytes10_map, "features/address_bytes10_map.fe"}
test_analysis! { abi_decode_complex, "features/abi_decode_complex.fe"}
test_analysis! { assert, "features/assert.fe"}
test_analysis! { associated_fns, "features/associated_fns.fe"}
test_analysis! { aug_assign, "features/aug_assign.fe"}
test_analysis! { base_tuple, "features/base_tuple.fe"}
test_analysis! { call_statement_with_args, "features/call_statement_with_args.fe"}
test_analysis! { call_statement_with_args_2, "features/call_statement_with_args_2.fe"}
test_analysis! { call_statement_without_args, "features/call_statement_without_args.fe"}
test_analysis! { checked_arithmetic, "features/checked_arithmetic.fe"}
test_analysis! { constructor, "features/constructor.fe"}
test_analysis! { create2_contract, "features/create2_contract.fe"}
test_analysis! { create_contract, "features/create_contract.fe"}
test_analysis! { create_contract_from_init, "features/create_contract_from_init.fe"}
test_analysis! { empty, "features/empty.fe"}
test_analysis! { events, "features/events.fe"}
test_analysis! { module_level_events, "features/module_level_events.fe"}
test_analysis! { external_contract, "features/external_contract.fe"}
test_analysis! { for_loop_with_break, "features/for_loop_with_break.fe"}
test_analysis! { for_loop_with_continue, "features/for_loop_with_continue.fe"}
test_analysis! { for_loop_with_static_array, "features/for_loop_with_static_array.fe"}
test_analysis! { if_statement, "features/if_statement.fe"}
test_analysis! { if_statement_2, "features/if_statement_2.fe"}
test_analysis! { if_statement_with_block_declaration, "features/if_statement_with_block_declaration.fe"}
test_analysis! { keccak, "features/keccak.fe"}
test_analysis! { math, "features/math.fe"}
test_analysis! { module_const, "features/module_const.fe"}
test_analysis! { multi_param, "features/multi_param.fe"}
test_analysis! { nested_map, "features/nested_map.fe"}
test_analysis! { numeric_sizes, "features/numeric_sizes.fe"}
test_analysis! { ownable, "features/ownable.fe"}
test_analysis! { pure_fn_standalone, "features/pure_fn_standalone.fe"}
test_analysis! { return_addition_i256, "features/return_addition_i256.fe"}
test_analysis! { return_addition_u128, "features/return_addition_u128.fe"}
test_analysis! { return_addition_u256, "features/return_addition_u256.fe"}
test_analysis! { return_array, "features/return_array.fe"}
test_analysis! { return_bitwiseand_u128, "features/return_bitwiseand_u128.fe"}
test_analysis! { return_bitwiseand_u256, "features/return_bitwiseand_u256.fe"}
test_analysis! { return_bitwiseor_u256, "features/return_bitwiseor_u256.fe"}
test_analysis! { return_bitwiseshl_u256, "features/return_bitwiseshl_u256.fe"}
test_analysis! { return_bitwiseshr_i256, "features/return_bitwiseshr_i256.fe"}
test_analysis! { return_bitwiseshr_u256, "features/return_bitwiseshr_u256.fe"}
test_analysis! { return_bitwisexor_u256, "features/return_bitwisexor_u256.fe"}
test_analysis! { return_bool_false, "features/return_bool_false.fe"}
test_analysis! { return_bool_inverted, "features/return_bool_inverted.fe"}
test_analysis! { return_bool_op_and, "features/return_bool_op_and.fe"}
test_analysis! { return_bool_op_or, "features/return_bool_op_or.fe"}
test_analysis! { return_bool_true, "features/return_bool_true.fe"}
test_analysis! { return_builtin_attributes, "features/return_builtin_attributes.fe"}
test_analysis! { return_complex_struct, "features/return_complex_struct.fe"}
test_analysis! { return_division_i256, "features/return_division_i256.fe"}
test_analysis! { return_division_u256, "features/return_division_u256.fe"}
test_analysis! { return_empty_tuple, "features/return_unit.fe"}
test_analysis! { return_eq_u256, "features/return_eq_u256.fe"}
test_analysis! { return_gt_i256, "features/return_gt_i256.fe"}
test_analysis! { return_gt_u256, "features/return_gt_u256.fe"}
test_analysis! { return_gte_i256, "features/return_gte_i256.fe"}
test_analysis! { return_gte_u256, "features/return_gte_u256.fe"}
test_analysis! { return_i128_cast, "features/return_i128_cast.fe"}
test_analysis! { return_i256, "features/return_i256.fe"}
test_analysis! { return_identity_u8, "features/return_identity_u8.fe"}
test_analysis! { return_identity_u16, "features/return_identity_u16.fe"}
test_analysis! { return_identity_u32, "features/return_identity_u32.fe"}
test_analysis! { return_identity_u64, "features/return_identity_u64.fe"}
test_analysis! { return_identity_u128, "features/return_identity_u128.fe"}
test_analysis! { return_identity_u256, "features/return_identity_u256.fe"}
test_analysis! { return_lt_i256, "features/return_lt_i256.fe"}
test_analysis! { return_lt_u128, "features/return_lt_u128.fe"}
test_analysis! { return_lt_u256, "features/return_lt_u256.fe"}
test_analysis! { return_lte_i256, "features/return_lte_i256.fe"}
test_analysis! { return_lte_u256, "features/return_lte_u256.fe"}
test_analysis! { return_mod_i256, "features/return_mod_i256.fe"}
test_analysis! { return_mod_u256, "features/return_mod_u256.fe"}
test_analysis! { return_msg_sig, "features/return_msg_sig.fe"}
test_analysis! { return_multiplication_i256, "features/return_multiplication_i256.fe"}
test_analysis! { return_multiplication_u256, "features/return_multiplication_u256.fe"}
test_analysis! { return_noteq_u256, "features/return_noteq_u256.fe"}
test_analysis! { return_pow_i256, "features/return_pow_i256.fe"}
test_analysis! { return_pow_u256, "features/return_pow_u256.fe"}
test_analysis! { return_subtraction_i256, "features/return_subtraction_i256.fe"}
test_analysis! { return_subtraction_u256, "features/return_subtraction_u256.fe"}
test_analysis! { return_u128_cast, "features/return_u128_cast.fe"}
test_analysis! { return_u256, "features/return_u256.fe"}
test_analysis! { return_u256_from_called_fn, "features/return_u256_from_called_fn.fe"}
test_analysis! { return_u256_from_called_fn_with_args, "features/return_u256_from_called_fn_with_args.fe"}
test_analysis! { revert, "features/revert.fe"}
test_analysis! { self_address, "features/self_address.fe"}
test_analysis! { send_value, "features/send_value.fe"}
test_analysis! { balances, "features/balances.fe"}
test_analysis! { sized_vals_in_sto, "features/sized_vals_in_sto.fe"}
test_analysis! { strings, "features/strings.fe"}
test_analysis! { structs, "features/structs.fe"}
test_analysis! { struct_fns, "features/struct_fns.fe"}
test_analysis! { ternary_expression, "features/ternary_expression.fe"}
test_analysis! { two_contracts, "features/two_contracts.fe"}
test_analysis! { u8_u8_map, "features/u8_u8_map.fe"}
test_analysis! { u16_u16_map, "features/u16_u16_map.fe"}
test_analysis! { u32_u32_map, "features/u32_u32_map.fe"}
test_analysis! { u64_u64_map, "features/u64_u64_map.fe"}
test_analysis! { u128_u128_map, "features/u128_u128_map.fe"}
test_analysis! { u256_u256_map, "features/u256_u256_map.fe"}
test_analysis! { while_loop, "features/while_loop.fe"}
test_analysis! { while_loop_with_break, "features/while_loop_with_break.fe"}
test_analysis! { while_loop_with_break_2, "features/while_loop_with_break_2.fe"}
test_analysis! { while_loop_with_continue, "features/while_loop_with_continue.fe"}
test_analysis! { abi_encoding_stress, "stress/abi_encoding_stress.fe"}
test_analysis! { data_copying_stress, "stress/data_copying_stress.fe"}
test_analysis! { tuple_stress, "stress/tuple_stress.fe"}
test_analysis! { type_aliases, "features/type_aliases.fe"}
test_analysis! { const_generics, "features/const_generics.fe" }
test_analysis! { const_local, "features/const_local.fe" }

test_analysis_ingot! { basic_ingot, "ingots/basic_ingot/src"}

fn build_snapshot(db: &dyn AnalyzerDb, module: items::ModuleId) -> String {
    let diagnostics = module
        .all_items(db)
        .iter()
        .flat_map(|item| match item {
            Item::Type(TypeDef::Alias(alias)) => vec![build_display_diagnostic(
                alias.data(db).ast.span,
                &alias.type_id(db).unwrap().display(db),
            )],
            Item::Type(TypeDef::Struct(struct_)) => [
                label_in_non_overlapping_groups(
                    db,
                    &struct_
                        .fields(db)
                        .values()
                        .map(|field| (field.data(db).ast.span, field.typ(db).unwrap()))
                        .collect::<Vec<_>>(),
                ),
                struct_
                    .functions(db)
                    .values()
                    .flat_map(|id| function_diagnostics(*id, db))
                    .collect(),
            ]
            .concat(),
            Item::Type(TypeDef::Contract(contract)) => [
                label_in_non_overlapping_groups(
                    db,
                    &contract
                        .fields(db)
                        .values()
                        .map(|field| (field.data(db).ast.span, field.typ(db).unwrap()))
                        .collect::<Vec<_>>(),
                ),
                contract
                    .events(db)
                    .values()
                    .flat_map(|id| event_diagnostics(*id, db))
                    .collect(),
                contract
                    .functions(db)
                    .values()
                    .flat_map(|id| function_diagnostics(*id, db))
                    .collect(),
            ]
            .concat(),
            Item::Trait(val) => val
                .all_functions(db)
                .iter()
                .map(|fun| {
                    build_display_diagnostic(fun.data(db).ast.span, &fun.signature(db).display(db))
                })
                .collect::<Vec<_>>(),
            Item::Impl(val) => val
                .all_functions(db)
                .iter()
                .map(|fun| {
                    build_display_diagnostic(fun.data(db).ast.span, &fun.signature(db).display(db))
                })
                .collect::<Vec<_>>(),
            Item::Function(id) => function_diagnostics(*id, db),
            Item::Constant(id) => vec![build_display_diagnostic(
                id.span(db),
                &id.typ(db).unwrap().display(db),
            )],
            Item::Event(id) => event_diagnostics(*id, db),
            // Built-in stuff
            Item::Type(TypeDef::Primitive(_))
            | Item::GenericType(_)
            | Item::BuiltinFunction(_)
            | Item::Intrinsic(_)
            | Item::Ingot(_)
            | Item::Module(_) => vec![],
        })
        .collect::<Vec<_>>();

    diagnostics_string(db.upcast(), &diagnostics)
}

fn new_diagnostic(labels: Vec<Label>) -> Diagnostic {
    Diagnostic {
        severity: Severity::Note,
        message: String::new(),
        labels,
        notes: vec![],
    }
}

fn label_in_non_overlapping_groups(
    db: &dyn AnalyzerDb,
    spans: &[(Span, impl Displayable)],
) -> Vec<Diagnostic> {
    // Accumulate labels in a vec until we reach a span that overlaps
    // the labeled range, then emit a Diagnostic with the accumulated labels
    // and begin again. This assumes that all spans are within the same file.
    let file_id = if let Some((span, _)) = spans.first() {
        span.file_id
    } else {
        return vec![];
    };

    spans
        .iter()
        .enumerate()
        .scan(
            (Span::zero(file_id), vec![]),
            |(labeled, labels), (idx, (span, attr))| {
                let mut diags = SmallVec::<[Diagnostic; 2]>::new();

                let overlaps = span.start < labeled.end && span.end > labeled.start;

                // If the current span overlaps with the union of the current set of labels,
                // emit a diagnostic, and clear the set of labels.
                if overlaps {
                    diags.push(new_diagnostic(labels.clone()));
                    labels.clear();
                    *labeled = *span;
                }
                labels.push(Label::primary(*span, attr.display(db).to_string()));
                *labeled += *span;

                // If this is the last thing to label, emit a diagnostic.
                if idx == spans.len() - 1 {
                    diags.push(new_diagnostic(labels.clone()));
                }
                Some(diags)
            },
        )
        .flatten()
        .collect()
}

fn function_diagnostics(fun: items::FunctionId, db: &dyn AnalyzerDb) -> Vec<Diagnostic> {
    let body = fun.body(db);
    [
        // signature
        vec![build_display_diagnostic(
            fun.data(db).ast.span,
            &fun.signature(db).display(db),
        )],
        // declarations
        label_in_non_overlapping_groups(db, &lookup_spans(&body.var_types, &body.spans)),
        // expressions
        label_in_non_overlapping_groups(db, &lookup_spans(&body.expressions, &body.spans)),
        // emits
        // TODO: log fully qualified path to event type
        //  eg foo::bar::MyEvent

        // CallType includes FunctionId,ContractId,StructId, so the debug output
        // may change when we add something to the std lib.
        // Disabling until we come up with a better label to use here.
        // label_in_non_overlapping_groups(db, &lookup_spans(&body.calls, &body.spans)),
    ]
    .concat()
}

fn event_diagnostics(event: items::EventId, db: &dyn AnalyzerDb) -> Vec<Diagnostic> {
    // Event field spans are a bit of a hassle right now
    label_in_non_overlapping_groups(
        db,
        &event
            .data(db)
            .ast
            .kind
            .fields
            .iter()
            .map(|node| node.span)
            .zip(
                event
                    .typ(db)
                    .fields
                    .iter()
                    .map(|field| field.typ.clone().unwrap()),
            )
            .collect::<Vec<(Span, TypeId)>>(),
    )
}

fn lookup_spans<T: Clone>(
    node_attrs: &IndexMap<NodeId, T>,
    spans: &HashMap<NodeId, Span>,
) -> Vec<(Span, T)> {
    node_attrs
        .iter()
        .map(|(id, attr)| (*spans.get(id).unwrap(), attr.clone()))
        .collect()
}

fn build_display_diagnostic<T: Display>(span: Span, attributes: &T) -> Diagnostic {
    let label = Label::primary(span, format!("{}", attributes));
    Diagnostic {
        severity: Severity::Note,
        message: String::new(),
        labels: vec![label],
        notes: vec![],
    }
}
