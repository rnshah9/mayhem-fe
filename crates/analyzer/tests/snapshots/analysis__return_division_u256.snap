---
source: crates/analyzer/tests/analysis.rs
expression: "build_snapshot(&db, module)"

---
note: 
  ┌─ return_division_u256.fe:2:5
  │  
2 │ ╭     pub fn bar(x: u256, y: u256) -> u256 {
3 │ │         return x / y
4 │ │     }
  │ ╰─────^ self: None, params: [{ label: None, name: x, typ: u256 }, { label: None, name: y, typ: u256 }] -> u256

note: 
  ┌─ return_division_u256.fe:3:16
  │
3 │         return x / y
  │                ^   ^ u256: Value
  │                │    
  │                u256: Value

note: 
  ┌─ return_division_u256.fe:3:16
  │
3 │         return x / y
  │                ^^^^^ u256: Value


